framework/src/tinyalsa/tinyalsa.c

```diff
@@ -497,15 +497,17 @@ int pcm_writei(struct pcm *pcm, const void *data, unsigned int frame_count)
 	}
 
 	pending = pcm_frames_to_bytes(pcm, frame_count);
-
 	while (pending > 0) {
-		nbytes = pending > pcm_frames_to_bytes(pcm, pcm->buffer_size) ? pcm_frames_to_bytes(pcm, pcm->buffer_size) : pending;
-		if (pcm->buf_idx < pcm->buffer_cnt) {
-			/* If we have empty buffers, fill them first */
-			memcpy(pcm->pBuffers[pcm->buf_idx]->samp, (char *)data + offset, nbytes);
-			apb = pcm->pBuffers[pcm->buf_idx];
-			pcm->buf_idx++;
-		} else {
+		bool found = false;
+		for (int i = 0; i < pcm->buffer_cnt; i++)
```

Comment:
I think we can make the logic a bit better, pcm->next_buf is not used in case of PCM_OUT type. How about using it instead?

I think it will be a bit faster too,


```
                if (pcm->next_buf != NULL) {
			apb = pcm->next_buf; --------->here
		if (pcm->buf_idx < pcm->buffer_cnt) {
			/* If we have empty buffers, fill them first */
			apb = pcm->pBuffers[pcm->buf_idx];
			pcm->next_buf = apb; ------------------>here
			pcm->buf_idx++;
		} else {
			/* We dont have any empty buffers. wait for deque message from kernel */
			size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
			if (size != sizeof(msg)) {
				/* Interrupted by a signal? What to do? */
				return oops(pcm, EINTR, "Interrupted while waiting for deque message from kernel\n");
			}
			if (msg.msgId == AUDIO_MSG_DEQUEUE) {
				apb = (struct ap_buffer_s *)msg.u.pPtr;
				pcm->next_buf = apb; ------------------> here
			} else if (msg.msgId == AUDIO_MSG_XRUN) {
				/* Underrun to be handled by client */
				return -EPIPE;
			} else {
				return oops(pcm, EINTR, "Recieved unexpected msg (id = %d) while waiting for deque message from kernel\n", msg.msgId);
			}
		}
```
and then pcm->next_buf can be enqueued in pcm_drain.